src/distributed_copy.c

```diff
@@ -0,0 +1,857 @@
+		i = 0;
+
+		foreach(shardIntervalCell, shardIntervalList)
+		{
+			ShardInterval *shardInterval = (ShardInterval *) lfirst(shardIntervalCell);
+			ShardId shardId = shardInterval->id;
+			LockShardData(shardId, ShareLock);
+			LockShardDistributionMetadata(shardId, ShareLock);
+			if (partitionType == HASH_PARTITION_TYPE)
+			{
+				int32 shardMinHashToken = INT32_MIN + (shardId * hashTokenIncrement);
+				int32 shardMaxHashToken = shardMinHashToken + (hashTokenIncrement - 1);
+				if (shardId == (shardCount - 1))
+				{
+					shardMaxHashToken = INT32_MAX;
+				}
+				if (DatumGetInt32(shardInterval->minValue) != shardMinHashToken ||
+					DatumGetInt32(shardInterval->maxValue) != shardMaxHashToken)
+				{
+					partitionType = RANGE_PARTITION_TYPE;
+				}
+			}
+		}
+
+		if (partitionType != HASH_PARTITION_TYPE)
+		{	
+			typeEntry = lookup_type_cache(intervalTypeId, TYPECACHE_CMP_PROC_FINFO);
+			compareFunction = &(typeEntry->cmp_proc_finfo);
+
+			i = 0;
+			foreach(shardIntervalCell, shardIntervalList)
+			{
+				shardIntervalCache[i++] = (ShardInterval *) lfirst(shardIntervalCell);
+			}
+			Assert(i == shardCount);
+			qsort_arg(shardIntervalCache, shardCount, sizeof(ShardInterval*), CompareTaskByMinHashToken, compareFunction);
+		}
+		while (true)
+		{
+			ShardInterval *shardInterval = NULL;
+			ShardId shardId = 0;
+			bool found = false;
+			MemoryContext oldContext = MemoryContextSwitchTo(tupleContext);
+			nextRowFound = NextCopyFrom(copyState, NULL, columnValues, columnNulls, NULL);
+			MemoryContextSwitchTo(oldContext);
+
+			if (!nextRowFound)
+			{
+				MemoryContextReset(tupleContext);
+				break;
+			}
+
+			/* write the row to the shard */
+			if (columnNulls[partitionColumn->varattno-1])
+			{
+				ereport(ERROR, (errcode(ERRCODE_NULL_VALUE_NOT_ALLOWED),
+								errmsg("cannot copy row with NULL value "
+									   "in partition column")));
+			}
+			partitionColumnValue = columnValues[partitionColumn->varattno-1];
+			if (partitionType == HASH_PARTITION_TYPE)
+			{
+				int hashedValue = DatumGetInt32(FunctionCall1(hashFunction,
+															  partitionColumnValue));
+				shardHashCode =
+					(int) ((uint32) (hashedValue - INT32_MIN) / hashTokenIncrement);
+				shardInterval = shardIntervalCache[shardHashCode];
+			}
+			else
+			{
+				int l = 0, r = shardCount;
+				while (l < r) 
+				{
+					int m = (l + r) >> 1;
+					if (DatumGetInt32(FunctionCall2Coll(compareFunction, 
+														DEFAULT_COLLATION_OID, 
+														partitionColumnValue, 
+														shardIntervalCache[m]->minValue)) < 0)
+					{
+						r = m;
+					} 
+					else if (DatumGetInt32(FunctionCall2Coll(compareFunction, 
+															 DEFAULT_COLLATION_OID, 
+															 partitionColumnValue, 
+															 shardIntervalCache[m]->maxValue)) <= 0)
+					{
+						shardInterval = shardIntervalCache[m];
+						break;
+					}
+					else
+					{
+						l = m + 1;
+					}
+				}
+			}
+			if (shardInterval == NULL)
+			{						
```

Comment:
I think we should now be able to find any existing shard in the cache, so if we don't we should just error out:

You can use the error message from https://github.com/citusdata/pg_shard/blob/develop/src/pg_shard.c#L738